File: authentication-service/src/test/java/com/microservice/authentication/web/util/CustomDefaultErrorAttributesTest.java
package com.microservice.authentication.web.util;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.boot.web.error.ErrorAttributeOptions;
import org.springframework.boot.web.servlet.error.DefaultErrorAttributes;
import org.springframework.http.HttpStatus;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.web.context.request.ServletWebRequest;
import org.springframework.web.server.ResponseStatusException;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(MockitoExtension.class)
class CustomDefaultErrorAttributesTest {

    @Test
    void testGetErrorAttributes() {
        CustomDefaultErrorAttributes customDefaultErrorAttributes = new CustomDefaultErrorAttributes();

        MockHttpServletRequest request = new MockHttpServletRequest();
        String name = DefaultErrorAttributes.class.getName() + ".ERROR";
        request.setAttribute(name, new RuntimeException("Connection refused!"));
        Map<String, Object> errorAttributes = customDefaultErrorAttributes.getErrorAttributes(new ServletWebRequest(request), ErrorAttributeOptions.defaults());

        assertThat(errorAttributes.get("status")).isEqualTo(503);
        assertThat(errorAttributes.get("timestamp")).isNotNull();
        assertThat(errorAttributes.get("error")).isNotNull();
        assertThat(errorAttributes.get("message").toString()).contains("Connection refused!");

        request.setAttribute(name, new BadCredentialsException("user is locked!"));
        errorAttributes = customDefaultErrorAttributes.getErrorAttributes(new ServletWebRequest(request), ErrorAttributeOptions.defaults());
        assertThat(errorAttributes.get("status")).isEqualTo(401);

        request.setAttribute(name, new ResponseStatusException(HttpStatus.NOT_FOUND));
        errorAttributes = customDefaultErrorAttributes.getErrorAttributes(new ServletWebRequest(request), ErrorAttributeOptions.defaults());
        assertThat(errorAttributes.get("status")).isEqualTo(404);
    }
}

File: kotlin-service/src/test/kotlin/com/microservice/kotlin/config/CustomDefaultErrorAttributesTest.kt
package com.microservice.kotlin.config

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.http.HttpStatus
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.web.server.ResponseStatusException

@ExtendWith(MockitoExtension::class)
internal class CustomDefaultErrorAttributesTest {

    @Test
    fun testGetErrorAttributes() {
        val customDefaultErrorAttributes = CustomDefaultErrorAttributes()

        val request = MockHttpServletRequest()
        var errorAttributes = customDefaultErrorAttributes.getErrorAttributes(request, RuntimeException("Connection refused!"), true)

        assertThat(errorAttributes["status"]).isEqualTo(503)
        assertThat(errorAttributes["timestamp"]).isNotNull()
        assertThat(errorAttributes["error"]).isNotNull()
        assertThat(errorAttributes["message"].toString()).contains("Connection refused!")

        errorAttributes = customDefaultErrorAttributes.getErrorAttributes(request, BadCredentialsException("user is locked!"), true)
        assertThat(errorAttributes["status"]).isEqualTo(401)

        errorAttributes = customDefaultErrorAttributes.getErrorAttributes(request, ResponseStatusException(HttpStatus.NOT_FOUND), true)
        assertThat(errorAttributes["status"]).isEqualTo(404)
    }
}
